### ISSUE TITLE
Tracking broken with Mongoid 6 and Rails 5

### ISSUE THREAD

**Issue by Startouf** [ankane/ahoy#227]
I just finished upgrading from Rails 5 beta to Rails 5.0.0.1, and without doing anything else I'm getting errors on automatic tracking

> Mongoid::Errors::InvalidValue
> Value of type ActionController::Parameters cannot be written to a field of type Hash

Here's the relevant part of the stack trace

```
mongoid (6.0.0) lib/mongoid/attributes.rb:333:in `validate_attribute_value'
mongoid (6.0.0) lib/mongoid/attributes.rb:168:in `block (2 levels) in write_attribute'
mongoid (6.0.0) lib/mongoid/threaded/lifecycle.rb:31:in `_assigning'
mongoid (6.0.0) lib/mongoid/attributes.rb:167:in `block in write_attribute'
mongoid (6.0.0) lib/mongoid/attributes/readonly.rb:36:in `as_writable_attribute!'
mongoid (6.0.0) lib/mongoid/attributes.rb:166:in `write_attribute'
mongoid (6.0.0) lib/mongoid/fields.rb:462:in `block (2 levels) in create_field_setter'
ahoy_matey (1.5.2) lib/ahoy/stores/mongoid_store.rb:28:in `block in track_event'
mongoid (6.0.0) lib/mongoid/document.rb:119:in `block (2 levels) in initialize'
mongoid (6.0.0) lib/mongoid/attributes/processing.rb:28:in `process_attributes'
mongoid (6.0.0) lib/mongoid/document.rb:118:in `block in initialize'
mongoid (6.0.0) lib/mongoid/threaded/lifecycle.rb:89:in `_building'
mongoid (6.0.0) lib/mongoid/document.rb:113:in `initialize'
ahoy_matey (1.5.2) lib/ahoy/stores/mongoid_store.rb:23:in `new'
ahoy_matey (1.5.2) lib/ahoy/stores/mongoid_store.rb:23:in `track_event'
ahoy_matey (1.5.2) lib/ahoy/tracker.rb:23:in `track'
ahoy_matey (1.5.2) app/controllers/ahoy/events_controller.rb:29:in `block in create'
ahoy_matey (1.5.2) app/controllers/ahoy/events_controller.rb:19:in `each'
ahoy_matey (1.5.2) app/controllers/ahoy/events_controller.rb:19:in `create'
```

**Comment by ankane** [comment]
Thanks for reporting @Startouf. Happy to accept a PR for this.

**Comment by Startouf** [comment]
I believe this is because Rails 5 uses new "ActionParameters" are not hashes for params. I believe a simple solution would be to convert them to hash manually (by simply callling to_h).

I'm not sure at what level the conversion should be done / where the parameters are indeed being written to the mongoid model.

We could convert the parameters directly from there ? https://github.com/ankane/ahoy/blob/master/app/controllers/ahoy/events_controller.rb by adding `to_h` ?

**Comment by Startouf** [comment]
Ah it's actually `to_unsafe_h` and not `to_h`

In the meantime I managed to monkeypatch it with

```
# https://github.com/ankane/ahoy/blob/master/app/controllers/ahoy/events_controller.rb
class EventsController < Ahoy::BaseController
    def create
      events =
        if params[:name]
          # legacy API
          params.blank? ? [params] : [params.to_unsafe_h]
        elsif params[:events]
          params[:events].blank? ? params[:events] : params.to_unsafe_h[:events]
        else
        ...
```
I probably won't have time to make a PR with a cleaner solution

**Comment by wa0x6e** [comment]
A very quick fix is to use `request.params` instead of just `params`, so that the returned value is a `HashWithIndifferentAccess` object instead of `ActionController::Parameters`.

This will bypass all the security benefits introduced by the new Parameters object.
Maybe we can add a new config option for the new params whitelist, and fallback to insecure hash only when that whitelist is empty, and trigger a deprecated notice in that case ?

**Comment by ankane** [comment]
Fixed with #229
